### PR TITLE
feat: refactor commits retrieval

### LINKED ISSUE #381: Bring changelog in-house
Replace our existing use of [changelog](https://npmjs.com/package/changelog) with internal logic.

Confusingly, `changelog` doesn't actually download *changelog files* (e.g. `CHANGELOG.md`) but instead just downloads all the commits from a repo between the npm publish dates of the from and to version. The logic we have now to download `CHANGELOG.md` is already "in house" and will be combined with new in-house logic for commit retrieval.

This new "get commits" function will take an npm package name, a "from" version and a "to" version. It should then:

1. Identify the source repository for the package from the npm metadata. It often needs some "massaging" but we already have a package that helps with that

2. Decide how we will identify the exact from/to commits. I think most npm releases these days actually list the commit hash so that sounds ideal (e.g. start at the commit after hash A and include up to hash B). Otherwise, we may need to fall back to timestamps (all commits after timestamp A and before timestamp B) but I'm not sure how common that is

3. Retrieve the commits! Ideally this should be done intelligently. It will undoubtedly need pagination (Github limits to maximum 100 per query) but we don't want to retrieve *all of them* in the entire repo

4. Filter/massage them and cache them (like we do today). Commits between two hashes or two timestamps should never change, so we may as well save ourselves extra GitHub queries.

The goal is that `changelog` can be removed from project dependencies while PRs still look essentially the same.

Notes:

- For now, we are concerned only with GitHub.com source packages, however in future we will want to support "internal" GitHub Enterprise ones as well as ones from GitLab. Do not try to solve everything at once but just keep it in mind

- This code will probably be combined together with the existing `CHANGELOG.md` logic as there will be a bit of overlap

### PR DESCRIPTION
**Pull request by kayoub5**
Refactor changelog logic to separate sources, and remove the `changelog` dependency.
Closes #381

### CODE REVIEW

**Comment by kayoub5** [comment]
@rarkins ready for review

File: lib/workers/pr/changelog/source-github.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+module.exports = {
+  getChangeLogJSON,
+};
+
+async function getTags(repository) {
+  try {
+    const versions = {};
+
+    const res = await ghGot(
+      `https://api.github.com/repos/${repository}/tags?per_page=100`
```
Comment (rarkins) [inline lib/workers/pr/changelog/source-github.js]: should be paginated

File: lib/workers/pr/changelog/source-cache.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+};
+
+function getCache(depName, fromVersion, newVersion) {
+  const tmpdir = process.env.RENOVATE_TMPDIR || os.tmpdir();
+  const cachePath = tmpdir + '/renovate-changelog-cache';
+  const cacheKey = `${depName}-${fromVersion}-${newVersion}`;
+  return [cachePath, cacheKey];
+}
+
+async function getChangeLogJSON(depName, fromVersion, newVersion) {
```
Comment (rarkins) [inline lib/workers/pr/changelog/source-cache.js]: no cache expiry?

File: lib/workers/pr/changelog/source-github.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    tags.forEach(tag => {
+      const version = isPinnedVersion(tag.name);
+      if (version) {
+        versions[version] = { gitHead: tag.name };
+      }
+    });
+    return versions;
+  } catch (err) {
+    logger.debug({ err, repository }, 'Failed to fetch Github tags');
```
Comment (rarkins) [inline lib/workers/pr/changelog/source-github.js]: should this usually happen? or be a WARN?

**Review by rarkins — DISMISSED** [review]
The logic/flow needs to be described somewhere, including caching logic

**Comment by felixfbecker** [comment]
I just stumbled across this PR in my search to try to configure Renovate for a specific package so that the `commitMessage` used is the commit message of the commit that is included in the release (if it only has one commit, which is usually the case with semantic-release). For this though I need access to the included commit message in the commitMessage handlebars template, which was possible before this PR, but now `changes` is always an empty array. Is there a way I could achieve this?

**Comment by rarkins** [comment]
@felixfbecker please raise a new issue maybe with a few more details of what you want to do

**Pull request merged**